### ISSUE TITLE
CMake support?

### ISSUE THREAD

**Issue by kangsheng89** [ConfettiFX/The-Forge#194]
is there any plan for offcial cmake support for the-forge?

https://github.com/DeanoC/The-Forge/blob/master/CMakeLists.txt

**Comment by cboiangiu** [comment]
I am currently trying to put together support for cmake with https://github.com/cboiangiu/The-Forge.cmake. I am using https://github.com/widberg/bgfx.cmake as a guide. Ideally I wish to build The-Forge as a package and then include the lib as a framework. 
Also working on https://github.com/cboiangiu/the-forge-starter to implement the lib with cmake support and draw a simple cube, while keeping the useful interfaces such as IApp (like they are used in the Unit Tests). The starter is based on https://github.com/pr0g/sdl-bgfx-imgui-starter.

**Comment by cboiangiu** [comment]
Also, in my opinion there should be a separation between the library itself and the IDEs used(or the project files themselves for that matter). As such, cmake support allows for easy integration with codebases and an easy way to get started.
When it comes to a product being developed with the library, the project file could be generated via cmake and then merged with some existing project file, that's maintained separately, that has the device specific configuration, bootstrapping etc.
That way you keep the existing testing in place and just separate the library from the project files, adding a layer in between that merges the cmake build with the project file for bootstrapping.

**Comment by kangsheng89** [comment]
Yes i totally agree with package it as lib. with cmake move, we also can consider using conan/vcpkg. By doing so, it will make the opensource community to adopt this framework

**Comment by mgood7123** [comment]
> I am currently trying to put together support for cmake with https://github.com/cboiangiu/The-Forge.cmake. I am using https://github.com/widberg/bgfx.cmake as a guide. Ideally I wish to build The-Forge as a package and then include the lib as a framework.
> Also working on https://github.com/cboiangiu/the-forge-starter to implement the lib with cmake support and draw a simple cube, while keeping the useful interfaces such as IApp (like they are used in the Unit Tests). The starter is based on https://github.com/pr0g/sdl-bgfx-imgui-starter.

this would also allow it to "just be used" with other IDE's that have support for cmake (or even simply via `cmake` itself like `make project` which would call cmake)

**Comment by THISISAGOODNAME** [comment]
https://github.com/CU-Production/The-Forge-CMake
You can build The-Forge first, and then use it as a library.

**Comment by Geenz** [comment]
https://github.com/Geenz/TheForgeTheCmake
This is my attempt at it.  It's fairly incomplete - but stuff like dynamic linking is there.  I'm working on adding unit tests and the like.

**Comment by JonahGoldsmith** [comment]
I have a working version for the newest update of The-Forge!
https://github.com/JonahGoldsmith/The-Forge-CMake
This Works on Windows Only but will work on MACOS and IOS in the next few days!
The Test Project Shows how to copy DLL's and Run the Shader Compiler just like the VS Projects Do!

**Comment by jochumdev** [comment]
I've made:

- [TheForgeCMakeTemplate](https://github.com/jochumdev/TheForgeCMakeTemplate)
- [TheForgeCMake](https://github.com/jochumdev/TheForgeCMake)

TheForgeCMake reflect's the UbuntuCodeLite/*.project files.

**Comment by Caellian** [comment]
I agree (subscribe).

I started reverse engineering Hades (for private use) a few days ago and started to work on adding CMake to make toggling support for different platforms easier.

I'll post a link with additions to The Forge.

In meantime, it would help if any of the current contributors could share whether there's any modifications that have been done to thirdparty libraries. I'll likely just use FetchContent for now, but it might be necessary to apply patches if there's changes.

Most other things should work as is, I'll probably just change imports so they're relative to modules as each target provides its own include paths.

**Comment by cboiangiu** [comment]
> I agree, but this has been previously discussed in [#158](https://github.com/ConfettiFX/The-Forge/issues/158) and the issue is that CMake supports too much so the tests wouldn't cover all the supported targets.
> 
> Not sure why simply checking `MSVC_VERSION` isn't an option as CMake could be made to throw errors if unsupported target is used. It's up to maintainers though and it seems they prefer using `.sln` files instead.
> 
> On a practical side of things, switching is almost certainly a cost that wouldn't be directly beneficial to them and if they're not familiar with CMake it just adds additional complexity to maintaining their `.sln`, so I don't believe they'd be willing to incur the cost of switching and **maintaining** even if someone provided a complete solution.

Wouldn't adding proper CMake support allow for easy generation of the current `.sln`s though?

**Comment by cboiangiu** [comment]
I do however support the idea of having a cross-platform `template` project similar to [this](https://github.com/ConfettiFX/The-Forge/issues/194#issuecomment-1407473878) with separate CMake support as a starting point. This would be best maintained separately from the current unit tests as to not disrupt the current pipelines and processes. Ideally this could be later merged here maybe, as it would be easier to maintain breaking changes or additions to the CMake project.

**Comment by Caellian** [comment]
> Wouldn't adding proper CMake support allow for easy generation of the current `.sln`s though? It's just that you focus on maintaining the CMake, instead, for better or for worse.

Yes and no. Differences start to show with complex configurations (e.g. game engines). A generator will never satisfy all requirements in all environments, which is why they don't want to use it. 

For the same reason, you might be better off setting up your own configuration with specialized dependencies, which I'm guessing they're doing for their clients.
- Many of used dependencies aren't what I'd personally pick when starting a new project for several different reasons (e.g. Lua vs LuaJIT, scheduling, etc.)

**Comment by sandroandrade** [comment]
I wouldn't invest any time in trying to add CMake support into Forge. I've started a PR a couple months ago with significant CMake support for all platforms (https://github.com/ConfettiFX/The-Forge/pull/310) but it was silently closed.

**Comment by tomwjerry** [comment]
While that is unfortunate, we can still try to do it like jochumdev et. al. and provide separate cmake files. It is not an optimal solution, but it works. Sadly I noticed that mine attempt did not work so I will keep looking at it.
